File: 2015/day3/solution.java
import java.io.*;
import java.util.*;

public class solution{

    public static int cntHouses(String moves) {
        Set<String> visited = new HashSet<>();
        int x = 0, y = 0;
        visited.add(x + "," + y); 

        for (char move : moves.toCharArray()) {
            switch (move) {
                case '^':
                    y++;
                    break;
                case 'v':
                    y--;
                    break;
                case '>':
                    x++;
                    break;
                case '<':
                    x--;
                    break;
            }
            visited.add(x + "," + y); // Adding the new position
        }

        return visited.size();
    }

    public static int cntHousesTwo(String moves) {
        Set<String> visited = new HashSet<>();
        int a = 0, b = 0; 
        int c = 0, d = 0; 
        visited.add(a + "," + b);

        for (int i = 0; i < moves.length(); i++) {
            char move = moves.charAt(i);
            if (i % 2 == 0) {
                switch (move) {
                    case '^':
                        b++;
                        break;
                    case 'v':
                        b--;
                        break;
                    case '>':
                        a++;
                        break;
                    case '<':
                        a--;
                        break;
                }
                visited.add(a + "," + b);
            } else {
                switch (move) {
                    case '^':
                        d++;
                        break;
                    case 'v':
                        d--;
                        break;
                    case '>':
                        c++;
                        break;
                    case '<':
                        c--;
                        break;
                }
                visited.add(c + "," + d); 
            }
        }

        return visited.size();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new FileReader("input.txt"));
        String moves = reader.readLine().trim();
        reader.close();

        System.out.println("Part 1: " + cntHouses(moves));
        System.out.println("Part 2: " + cntHousesTwo(moves));
    }
}

File: 2015/day3/solution.py
def cnt_houses(moves):
    visited = set()
    x,y = 0,0
    visited.add((x,y))

    for move in moves:
        if move == '^':
            y += 1
        elif move == 'v':
            y -= 1
        elif move == '>':
            x += 1
        elif move == '<':
            x -= 1
        visited.add((x,y))
    return len(visited)


def cnt_houses_two(moves):
    visited = set()
    a,b = 0,0
    c,d = 0,0
    visited.add((a,b))

    for i,move in enumerate(moves):
        if i%2 == 0:
            if move == '^':
                b += 1
            elif move == 'v':
                b -= 1
            elif move == '>':
                a += 1
            elif move == '<':
                a -= 1
            visited.add((a,b))
        else:
            if move == '^':
                d += 1
            elif move == 'v':
                d -= 1
            elif move == '>':
                c += 1
            elif move == '<':
                c -= 1
            visited.add((c,d))
    return len(visited)


            


with open('input.txt', "r") as file:
    moves = file.read().strip()

print("Part 1: ", cnt_houses(moves))
print("Part 2: ", cnt_houses_two(moves))
